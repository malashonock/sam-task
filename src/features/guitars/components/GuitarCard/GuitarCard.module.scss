.card {
  width: 300px;
  min-width: 300px;
  transition: 0.25s ease;

  &:hover {
    filter: drop-shadow(0 0 3px grey);
    transition: 0.25s ease;
  }
}