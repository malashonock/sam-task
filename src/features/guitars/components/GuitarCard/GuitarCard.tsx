import { useCallback, useContext } from 'react';
import ShoppingCartContext, {
  ShoppingCartState,
} from 'features/shoppingCart/model/ShoppingCartContext';
import Guitar, { areGuitarsEqual } from '../../model/Guitar';
import styles from './GuitarCard.module.scss';

interface GuitarCardProps {
  guitar: Guitar;
}

export default function GuitarCard({ guitar }: GuitarCardProps): JSX.Element {
  const { model, price, imageFilename }: Guitar = guitar;

  const { shoppingCart, dispatchShoppingCartAction }: ShoppingCartState = useContext(
    ShoppingCartContext
  ) as ShoppingCartState;

  const isInCart = useCallback((): boolean => {
    return shoppingCart.guitars.findIndex((itemInCart) => areGuitarsEqual(itemInCart, guitar)) > -1;
  }, [guitar, shoppingCart]);

  const handleShoppingCartToggle = useCallback((): void => {
    if (!isInCart()) {
      dispatchShoppingCartAction({ type: 'add', item: guitar });
    } else {
      dispatchShoppingCartAction({ type: 'remove', item: guitar });
    }
  }, [dispatchShoppingCartAction, guitar, isInCart]);

  return (
    <div className={['card', styles.card].join(' ')}>
      <div className="card-img p-3">
        <img src={`./img/${imageFilename}`} className="card-img-top img-fluid" alt={model} />
      </div>
      <div className="card-body text-center">
        <h6 className="card-title">{model}</h6>
        <h2 className="card-title">{`${price.toLocaleString('en-US', {
          style: 'currency',
          currency: 'USD',
        })}`}</h2>
        <button
          type="button"
          className={`btn btn-${isInCart() ? 'danger' : 'primary'} m-2`}
          onClick={handleShoppingCartToggle}
        >
          {isInCart() ? 'Remove from cart' : 'Add to cart'}
        </button>
      </div>
    </div>
  );
}
