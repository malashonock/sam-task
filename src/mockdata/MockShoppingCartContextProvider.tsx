import { PropsWithChildren, useReducer } from 'react';
import ShoppingCart from '../features/shoppingCart/model/ShoppingCart';
import ShoppingCartContext from '../features/shoppingCart/model/ShoppingCartContext';
import { shoppingCartReducer } from '../features/shoppingCart/model/ShoppingCartReducer';

export default function MockShoppingCartContextProvider({
  mockShoppingCart,
  children,
}: { mockShoppingCart: ShoppingCart } & PropsWithChildren): JSX.Element {
  const [shoppingCart, dispatchShoppingCartAction] = useReducer(
    shoppingCartReducer,
    mockShoppingCart
  );

  return (
    <ShoppingCartContext.Provider value={{ shoppingCart, dispatchShoppingCartAction }}>
      {children}
    </ShoppingCartContext.Provider>
  );
}
