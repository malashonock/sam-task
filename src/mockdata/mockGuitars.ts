import Guitar from 'features/guitars/model/Guitar';

const guitar1: Guitar = {
  category: 'Electric guitars',
  brand: 'Gibson',
  model: 'Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry',
  description:
    'Semi-hollowbody Electric Guitar with Maple/Poplar Body, Mahogany Neck, Rosewood Fingerboard, and 2 Humbucking Pickups - Sixties Cherry',
  color: 'cherry',
  price: 3499,
  inStock: true,
  rating: 5,
  popularity: 8,
  imageFilename: 'Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry.webp',
};

const guitar2: Guitar = {
  category: 'Electric guitars',
  brand: 'ESP',
  model: 'ESP LTD James Hetfield Signature Snakebyte - Snow White',
  description:
    'Solidbody Electric Guitar with Mahogany Body, Mahogany Neck, Ebony Fingerboard, and 2 Active EMG Humbucking Pickups - Snow White',
  color: 'white',
  price: 1499,
  inStock: false,
  rating: 5,
  popularity: 12,
  imageFilename: 'ESP LTD James Hetfield Signature Snakebyte - Snow White.webp',
};

const guitar3: Guitar = {
  category: 'Acoustic guitars',
  brand: 'Breedlove',
  model: 'Breedlove Oregon Concert CE Acoustic-Electric Guitar - Natural Myrtlewood',
  description:
    '6-string Acoustic-electric Guitar with Myrtlewood Top, Back, and Sides, Maple Neck, Ebony Fingerboard, and LR Baggs Element Electronics - Natural',
  color: 'natural',
  price: 2499,
  inStock: true,
  rating: 5,
  popularity: 2,
  imageFilename: 'Breedlove Oregon Concert CE Acoustic-Electric Guitar - Natural Myrtlewood.webp',
};

const mockGuitars: { [name: string]: Guitar[] } = {
  none: [],

  oneGuitar: [guitar1],

  severalGuitars: [guitar1, guitar2, guitar3],

  manyGuitars: new Array<Guitar>(100).fill(guitar1).map((guitar, index) => {
    return {
      ...guitar,
      model: `${guitar.model} (${index + 1})`,
    };
  }),
};

export default mockGuitars;
