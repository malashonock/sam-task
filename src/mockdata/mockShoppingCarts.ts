import ShoppingCart from '../features/shoppingCart/model/ShoppingCart';
import mockGuitars from './mockGuitars';

const mockShoppingCarts: { [name: string]: ShoppingCart } = {
  empty: {
    guitars: [],
    pendingOverflow: false,
  },

  oneGuitarNoOverflow: {
    guitars: mockGuitars.oneGuitar,
    pendingOverflow: false,
  },

  oneGuitarPendingOverflow: {
    guitars: mockGuitars.oneGuitar,
    pendingOverflow: true,
  },
};

export default mockShoppingCarts;
