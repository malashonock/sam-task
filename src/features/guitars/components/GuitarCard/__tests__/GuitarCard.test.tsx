import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import GuitarCard from '../GuitarCard';
import mockGuitars from 'mockdata/mockGuitars';
import MockShoppingCartContextProvider from 'mockdata/MockShoppingCartContextProvider';
import mockShoppingCarts from 'mockdata/mockShoppingCarts';

describe('GuitarCard component', () => {
  it('matches snapshot', () => {
    const guitarCard = renderer.create(
      <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.empty}>
        <GuitarCard guitar={mockGuitars.oneGuitar[0]} />
      </MockShoppingCartContextProvider>
    );

    expect(guitarCard).toMatchSnapshot();
  });

  describe('Add to/remove from cart button', () => {
    describe('given that guitar prop is NOT in the shopping cart', () => {
      it('has class "bnt-primary"', () => {
        const { container } = render(
          <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.empty}>
            <GuitarCard guitar={mockGuitars.oneGuitar[0]} />
          </MockShoppingCartContextProvider>
        );

        const button = container.querySelector('.btn');
        expect(button).toHaveClass('btn-primary');
      });

      it('has text "Add to cart"', () => {
        const { container } = render(
          <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.empty}>
            <GuitarCard guitar={mockGuitars.oneGuitar[0]} />
          </MockShoppingCartContextProvider>
        );

        const button = container.querySelector('.btn');
        expect(button).toHaveTextContent('Add to cart');
      });
    });

    describe('given that guitar prop IS in the shopping cart', () => {
      it('has class "btn-warning"', () => {
        const { container } = render(
          <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.oneGuitarNoOverflow}>
            <GuitarCard guitar={mockGuitars.oneGuitar[0]} />
          </MockShoppingCartContextProvider>
        );

        const button = container.querySelector('.btn');
        expect(button).toHaveClass('btn-warning');
      });

      it('has text "Add one more"', () => {
        const { container } = render(
          <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.oneGuitarNoOverflow}>
            <GuitarCard guitar={mockGuitars.oneGuitar[0]} />
          </MockShoppingCartContextProvider>
        );

        const button = container.querySelector('.btn');
        expect(button).toHaveTextContent('Add one more');
      });
    });
  });
});
