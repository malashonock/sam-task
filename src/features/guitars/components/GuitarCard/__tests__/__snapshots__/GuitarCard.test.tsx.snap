// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuitarCard component matches snapshot 1`] = `
<div
  className="card card"
>
  <div
    className="card-img p-3"
  >
    <img
      alt="Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry"
      className="card-img-top img-fluid"
      src="./img/Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry.webp"
    />
  </div>
  <div
    className="card-body text-center"
  >
    <h6
      className="card-title"
    >
      Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry
    </h6>
    <h2
      className="card-title"
    >
      $3,499.00
    </h2>
    <button
      className="btn btn-primary m-2"
      onClick={[Function]}
      type="button"
    >
      Add to cart
    </button>
  </div>
</div>
`;

exports[`GuitarCard component renders title 1`] = `
<div
  className="card card"
>
  <div
    className="card-img p-3"
  >
    <img
      alt="Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry"
      className="card-img-top img-fluid"
      src="./img/Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry.webp"
    />
  </div>
  <div
    className="card-body text-center"
  >
    <h6
      className="card-title"
    >
      Gibson ES-335 Semi-hollowbody Electric Guitar - Sixties Cherry
    </h6>
    <h2
      className="card-title"
    >
      $3,499.00
    </h2>
    <button
      className="btn btn-primary m-2"
      onClick={[Function]}
      type="button"
    >
      Add to cart
    </button>
  </div>
</div>
`;
